drivers/hmis_csv_importer/app/jobs/hmis_csv_importer/cleanup/expire_base_job.rb

```diff
@@ -0,0 +1,138 @@
+      end
+
+      models_to_run.each do |model|
+        with_lock(model) do
+          benchmark(model.table_name.to_s) do
+            process_model(model) if sufficient_imports?
+          end
+        end
+      end
+    end
+
+    def with_lock(model)
+      lock_name = "#{self.class.name.demodulize}:#{model.table_name}"
+      log_model.with_advisory_lock(lock_name, timeout_seconds: 0) do
+        yield
+      end
+    end
+
+    def process_model(model)
+      start_time = Time.current
+      # TODO: this can be removed (or at least the count query could be removed once we're comfortable with the results)
+      log "Start Processing: #{model.table_name}, rows overall: #{model.with_deleted.count}"
+
+      with_tmp_table(model) do |table_name|
+        populate_tmp_table(model, table_name)
+        write_expirations(model, table_name)
+      end
+
+      # TODO: this can be removed (or at least the count query could be removed once we're comfortable with the results)
+      log "Completed Processing: #{model.table_name}, rows expired: #{model.with_deleted.where(expired: true).count} in #{elapsed_time(Time.current - start_time)}"
+    end
+
+    private def with_tmp_table(model)
+      prefix = model.table_name.downcase
+      table_name = "#{prefix}_tmp_exp"
+      model.connection.create_table table_name, id: false, temporary: true do |t|
+        t.references :source, null: false, index: false
+      end
+      yield(table_name)
+    ensure
+      # drop tmp table explicitly since we will continue processing additional tables with this session
+      model.connection.drop_table table_name
+    end
+
+    private def write_expirations(model, tmp_table_name)
+      model.connection.execute(expiration_update_query(model, tmp_table_name))
+    end
+
+    # this could be batched if needed
+    private def expiration_update_query(model, tmp_table_name)
+      <<~SQL
+        UPDATE #{model.quoted_table_name} source_table
+        SET expired = true
+        FROM #{tmp_table_name} tmp_table
+        WHERE tmp_table.source_id = source_table.id
+        AND #{log_id_field} < #{min_age_protected_id}
+      SQL
+    end
+
+    private def populate_tmp_table(model, tmp_table_name)
+      model.connection.execute(populate_tmp_table_query(model, tmp_table_name))
+    end
+
+    private def populate_tmp_table_query(model, tmp_table_name)
+      <<~SQL
+        INSERT INTO #{tmp_table_name} (source_id)
+        SELECT id FROM (#{relevant_id_query(model)}) as relevant_ids
+      SQL
+    end
+
+    private def relevant_id_query(model)
+      key_field = model.hud_key
+      <<~SQL
+        SELECT id, #{log_id_field} FROM (
+          SELECT id, #{log_id_field}, row_number() OVER (
+            PARTITION BY "#{key_field}", data_source_id ORDER BY id DESC
+          ) AS row_num
+          FROM #{model.quoted_table_name}
+        ) subquery
+        WHERE subquery.row_num > #{@retain_item_count}
```

Comment:
What do you think about adding an `AND subquery.expired != true` here, and the associated select of the `expired` column.  That would allow this to be resumable and would cut down significantly on writes, which is one of the issues we're seeing with the burst balance in RDS.  Sadly, the update being in a transaction means those writes aren't recorded until the transaction commits.  Is there any way we can take that out of the transaction?  Then it really would be resumable, and I don't think there's any reason it would be a problem if we marked things expired in some incomplete way.

Reply:
OH!  Or maybe since we have the temp table (if we add back in the ID column) we could easily batch the update with some simple pagination.

Reply:
Great, thanks for adding that!